languages/python/sqlalchemy-oso/sqlalchemy_oso/partial.py

```diff
@@ -0,0 +1,171 @@
+    else:
+        path = dot_op_path(right)
+        assert path
+        value = left
+
+    path, field_name = path[:-1], path[-1]
+    return translate_dot_op(
+        path, session, model, functools.partial(emit_compare, field_name, value)
+    )
+
+
+def translate_in(expression, session, model):
+    assert expression.operator == "In"
+    left = expression.args[0]
+    right = expression.args[1]
+
+    # IN means at least something must be contained in the property.
+
+    # There are two possible types of in operations. In both, the right hand side
+    # should be a dot op.
+
+    # Partial In: LHS is an expression
+    if isinstance(left, Expression):
+        path = dot_op_path(right)
+        assert path
+
+        return translate_dot_op(
+            path, session, model, functools.partial(emit_subexpression, left)
+        )
+    else:
+        # Contains: LHS is not an expression.
+        # TODO (dhatch) Missing check, left type must match type of the target?
+        path = dot_op_path(right)
+        assert path
+        path, field_name = path[:-1], path[-1]
+        return translate_dot_op(
+            path, session, model, functools.partial(emit_contains, field_name, left)
+        )
+
+
+def translate_dot_op(path: List[str], session: Session, model, func: EmitFunction):
+    if len(path) == 0:
+        return func(session, model)
+    else:
+        property, model, is_multi_valued = get_relationship(model, path[0])
+        if not is_multi_valued:
+            return property.has(translate_dot_op(path[1:], session, model, func))
+        else:
+            return property.any(translate_dot_op(path[1:], session, model, func))
+
+
+def get_relationship(model, field_name: str):
+    """Get the property object for field on model. field must be a relationship field.
+
+    :returns: (property, model, is_multi_valued)
+    """
+    property = getattr(model, field_name)
+    assert isinstance(property.property, RelationshipProperty)
+    relationship = property.property
+    model = property.entity.class_
+
+    return (property, model, relationship.uselist)
+
+
+def emit_compare(field_name, value, session, model):
+    """Emit a comparison operation comparing the value of ``field_name`` on ``model`` to ``value``."""
+    property = getattr(model, field_name)
+    return property == value
+
+
+def emit_subexpression(sub_expression: Expression, session: Session, model):
+    """Emit a sub-expression on ``model``."""
+    return translate_expr(sub_expression, session, model)
+
+
+def emit_contains(field_name, value, session, model):
+    """Emit a contains operation, checking that multi-valued relationship field ``field_name`` contains ``value``."""
+    # TODO (dhatch): Could this be valid for fields that are not relationship fields?
+    property, model, is_multi_valued = get_relationship(model, field_name)
+    assert is_multi_valued
+
+    return property.contains(value)
+
+
+# TODO (dhatch): Move this helper into base.
```

Comment:
and the test in `languages/python/sqlalchemy-oso/tests/test_partial.py`

Reply:
yeah

Reply:
noted down